### DISCUSSION TITLE
The appropriate usage of transactions in Prisma

### DISCUSSION THREAD

**Discussion by kim-sung-jee** [prisma/prisma#21201] [Q&A] [answered]
### Question

**The following Prisma service exists.**

```
@Injectable()
export class PrismaService extends PrismaClient implements OnModuleInit {
  async onModuleInit() {
    await this.$connect();
  }
}
```

**Here is the following transaction process.**

`A transaction for creating a user:`
1. Transaction Begin (await this.prismaService.$transaction((tx) => {}))
2. tx.user.create => (userId is 11)
3. tx.user.find**(where: userId == 11)
4. Transaction End

My question is as follows: 
If the purpose of step 3 is to retrieve data that is currently reflected in the actual database, shouldn't step 3 use `this.prismaService` instead of `tx`? When querying with `tx`, it retrieves both the transaction context and the records reflected in the actual database (always returning 1 or more). But when querying with `prisma.service`, it only retrieves records that are actually reflected in the database. Therefore, I'm curious if using `prisma.service` would be a more natural behavior.

**Is there any potential issue with executing this.prismaService.user.find\*\*() within a transaction (tx)?**

**Comment by ludralph** [comment, marked as answer]
Hi @kim-sung-jee 👋 

Thank you for raising this question.

Using `this.prismaService.user.find` within a transaction is not necessarily wrong or problematic, but it serves a different purpose. It retrieves data from the actual database, not considering any changes made within the transaction. The reason is that Prisma's $transaction API is designed to handle multiple database operations as a single unit of work. When you use `tx` to perform operations, Prisma ensures that either all operations are successful, or if any operation fails, all operations are rolled back, maintaining the integrity of your data.

In the context of a transaction, `tx.user.find**()` is used to query data within the transaction scope. This means it can see the changes made within the same transaction before those changes are committed to the database. On the other hand, `this.prismaService.user.find**()` operates outside the transaction scope. It queries the database directly and can only see the changes that have been committed to the database. Therefore, if you use `this.prismaService.user.find**()` within a transaction, it may not see the changes made within the same transaction until those changes are committed.

As for potential issues, using `this.prismaService.user.find**()` within a transaction could lead to inconsistent results because it operates outside the transaction scope. It might not see the latest changes made within the same transaction. In conclusion, if you want to query data that reflects the changes made within the same transaction, you should use `tx.user.find**()`. If you want to query data that reflects the changes committed to the database, you should use `this.prismaService.user.find**()`, but be aware that it operates outside the transaction scope and may not see the latest changes made within the same transaction.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏
